### DISCUSSION TITLE
Inquiry on Inputting Measured Wind and Wave Data into OpenFAST v3.0.0

### DISCUSSION THREAD

**Discussion by niay000** [OpenFAST/openfast#2467] [Q&A]
Dear All,

I hope this message finds you well.

I have a couple of questions regarding the use of OpenFAST version 3.0.0 for my work:

Inputting Measured Wind Speed Data at Multiple Heights:
I have time-varying measured wind speed data (t, x, y, z) collected at multiple heights, such as hub height and tower base position. I would like to input this data into OpenFAST v3.0.0 for simulation purposes.

Is this possible with OpenFAST v3.0.0?
If so, what format should the data files be prepared in?
What specific changes or configurations are required in the input files to accommodate this data?(ex;inflowwin.dat?,etc...?)

Inputting Measured Wave Data:
Similarly, I have measured wave data representing time-varying water levels (t, eta) that I wish to input into OpenFAST v3.0.0.

Is it possible to incorporate this wave data into OpenFAST v3.0.0?
If so, what format should the wave data files be in?
What modifications or settings need to be adjusted in the input files to integrate this wave data effectively?(ex; hydrodyn.dat?,etc...?)

Your guidance on these matters would be greatly appreciated, as it will significantly aid in the accuracy and reliability of my simulations.

Thank you very much for your time and assistance.

Best regards,

**Comment by jjonkman** [comment]
Dear @niay000,

Yes, OpenFAST can make use of user-specified wind and wave data time series.

For user-specified wind data, presumably you want to prescribe wind time series at one or more points, and generate synthetic turbulence at other grid points that is correlated with the known times series.  For this, you can use the `TurbModel` = `TIMESR` option of TurbSim.  The use of PyConTurb from DTU is another option, which is more effective if you have multiple points where the time series are known.

For user-specified wave data, you can use the `WaveMod` = 5 option of HydroDyn (an input that will move to the SeaState module in the upcoming release of OpenFAST v4.0).

You can find examples and documentation of both of these options in the repository.

Best regards,

**Reply by niay000** [reply]
Dear jjonkman,

Thank you for your response. It was very helpful.
 I will check the repository as suggested.
 If I have any further questions, I would like to ask again.

Best regards,

**Reply by niay000** [reply]
Dear　jjomkman,

I apologize for any misunderstanding; it seems I did not convey my question clearly. My intention was to load the time series data directly. To achieve this, I have used the InflowWind.dat file.

For the wind speed time series data in the InflowWind.dat file, I set WindType = 2 and provided the appropriate filename for the Filename_Uni parameter. By doing this, I was able to achieve the desired result.
[sample_inflowwinddat.txt](https://github.com/user-attachments/files/17596459/sample_inflowwinddat.txt)
[inflow_uni_test.txt](https://github.com/user-attachments/files/17596465/inflow_uni_test.txt)


I now have an additional question:

①Can multiple height time series data for the x-direction be loaded in this setup? For example, I would like to load different time series data for heights of 150.0 m and 75.0 m.


Regarding the wave height time series data in the HydroDyn.dat file, I have set WaveMod = 5 and specified the WvKinFile with the correct filename (excluding the file extension). Additionally, I set WvDiffQTF and WvSumQTF (2ND-ORDER WAVES) to False, and WvLowCOff to 0.0. However, I noticed that the output wave height data appears at a lower position than the input data.

Attached, you will find an image (u_eta_comp.png) and a sample of the HydroDyn.dat file (sample_hydrodyn.txt) for reference.
![u_eta_comp](https://github.com/user-attachments/assets/54714177-5a56-4646-8f3a-c8cadb8eb09e)

[sample_hydrodyn.txt](https://github.com/user-attachments/files/17596442/sample_hydrodyn.txt)

②Could you please help me understand the reasons for this discrepancy and advise on how I can ensure that the output matches the input data correctly?
Thank you for your assistance, and I look forward to your response.

**Reply by jjonkman** [reply]
Dear @niay000,

Regarding support for different wind time series, that is what the `TIMESR` option of TurbSim can support.  The uniform wind files, per their name, can only support uniform wind data.

Regarding `WaveMod` = 5, the difference I see is that the mean is different.  HydroDyn requires that the mean of the wave-elevation time series be zero, and if the user-specified wave-elevation time series has a nonzero mean, HydroDyn will remove it.

Best regards,

**Reply by niay000** [reply]
Dear @jjonkman,

Thank you very much for your guidance on the settings for TurbSim and HydroDyn.

I now understand that TurbSim is indeed well-suited for reading wind data at multiple heights. I have attempted to input data where wind speed gradually increases while fluctuating over time. However, the generated bts file appears to have averaged out the wind speed variations, losing the progressive increase that was present in the original data.

![test2](https://github.com/user-attachments/assets/fa713ee1-5f5c-4edb-a632-5db9fdc00f7a)


Is it possible to bypass the re-calculation and retain the original wind speed variations directly in the bts file?

[15mw_test2_TimeSRinp.txt](https://github.com/user-attachments/files/17625513/15mw_test2_TimeSRinp.txt)

[Test2_timeSR.txt](https://github.com/user-attachments/files/17625515/Test2_timeSR.txt)

I also greatly appreciate the clarification on the HydroDyn settings. Thank you once again for your assistance.

Kind regards,

**Reply by jjonkman** [reply]
Dear @niay000,

TurbSim generally assumes that turbulence is represented as turbulent fluctuations superimposed on a mean velocity profile.  As such, TurbSim will remove the mean from the time series and compute the spectral amplitudes and phases (via FFT from the zero-mean resultant).  Given that your signal is more like a sine wave superimposed atop a linear ramp, my guess is the sine wave can be well represented by the FFT, but the linear ramp cannot, resulting in a mismatch between your user-specified time series and the resulting output.

Best regards,

**Reply by niay000** [reply]
Dear @jjonkman,

Thank you very much for your previous response. I would like to kindly ask for further clarification regarding the input of wind speed data into OpenFAST.

We are observing time-series data from a monitoring tower at three different heights: hub height, the midpoint between the tower base and top, and the tower base. The data includes two horizontal components of wind speed, recorded at 0.5-second intervals, with no vertical component. We are considering interpolating the wind speeds between these observation points for use in OpenFAST.

Could you please confirm if it is possible to input this type of time-series data directly into OpenFAST? If so, we would greatly appreciate your advice on the best method for formatting and preparing the data accordingly.

Thank you again for your assistance, and I look forward to your response.

Best regards,

**Reply by jjonkman** [reply]
Dear @niay000,

Yes, the `TIMESR` feature of TurbSim supports user-specified time series at multiple points, with the limitation that only one of the points can be used as a reference for spatial coherence to other points in the domain added within TurbSim.  If you want spatial coherence based on all user-specified time series, the PyConTurb software from DTU is a preferred alternative to TurbSim.

I'm not sure I understand why you'd want to interpolate the data between points, as interpolation will smooth out the turbulence a bit.  It would be better to use the raw data at the points directly.

FYI: a 0.5-s interval for inflow is a bit coarse for aero-elastic simulations, which may impact the accuracy of higher-frequency response and the resulting fatigue loads predicted by OpenFAST.

Best regards,

**Reply by niay000** [reply]
Dear @jjonkman ,

Thank you very much for your prompt response. Your explanation was very clear, and I now have a much better understanding of the matter.

As I continue to work through the details, I may have additional questions. If so, I would greatly appreciate the opportunity to reach out again for further clarification.

Best regards,
